### ISSUE TITLE
New step model rejects attempts to use the same step on multiple records

### ISSUE THREAD

**Issue by akoshelev** [private-attribution/ipa#135]
After we landed #122 , our previous model of interaction between steps no longer works. We thought that using finite number of steps that define the circuit and executing this circuit for many records is optimal, but debug assertions added in the mentioned PR make it no longer possible.

Demo program to demonstrate the issue

```rust
#[cfg(test)]
mod toy_demo {
    use crate::field::{Field, Fp31};
    use crate::helpers::fabric::Network;
    use crate::protocol::context::ProtocolContext;
    use crate::protocol::{QueryId, RecordId, Step};
    use crate::secret_sharing::Replicated;
    use futures::StreamExt;
    use futures_util::stream::FuturesOrdered;
    use rand::thread_rng;
    use crate::test_fixture::{make_contexts, make_world, share};


    /// Takes a collection as input and executes `MultiplyTwice` protocol on every input row.
    #[derive(Debug)]
    struct ToyProtocol<'a, F> {
        input: &'a [(Replicated<F>, Replicated<F>)]
    }

    /// This protocol runs per row and just multiplies a*b*a
    struct MultiplyTwice<F> {
        a: Replicated<F>,
        b: Replicated<F>,
        record_id: RecordId
    }

    enum ToyStep {
        Step1,
        Step2
    }

    impl <'a, F: Field> ToyProtocol<'a, F> {
        async fn execute<N: Network>(self, ctx: &ProtocolContext<'a, N>) -> Vec<Replicated<F>> {

            let mut futures = FuturesOrdered::new();

            // for every input row, execute `MultiplyTwice`
            for (record_id, r) in self.input.iter().enumerate() {
                let record_id = RecordId::from(record_id as u32);
                let sub_protocol = MultiplyTwice { a: r.0, b: r.1, record_id };
                futures.push_back(sub_protocol.execute(&ctx));
            }

            let res = futures.collect::<Vec<_>>().await;
            res
        }
    }

    impl <F: Field> MultiplyTwice<F> {
        async fn execute<N: Network>(self, ctx: &ProtocolContext<'_, N>) -> Replicated<F> {
            // error occurs here because we narrow the same context more than once (once per record)
            let ctx = ctx.narrow(&ToyStep::Step1);
            let r = ctx.multiply(self.record_id).await.execute(self.a, self.b).await.unwrap();

            let ctx = ctx.narrow(&ToyStep::Step2);
            ctx.multiply(self.record_id).await.execute(self.a, r).await.unwrap()
        }
    }

    impl Step for ToyStep {}

    impl AsRef<str> for ToyStep {
        fn as_ref(&self) -> &str {
            match self {
                ToyStep::Step1 => "toy_step_1",
                ToyStep::Step2 => "toy_step_2",
            }
        }
    }

    #[tokio::test]
    pub async fn test() {
        let world = make_world(QueryId);
        let contexts = make_contexts(&world);
        let mut rng = thread_rng();
        let data = vec![
            (Fp31::from(10u128), Fp31::from(2u128)),
            (Fp31::from(5u128), Fp31::from(3u128)),
        ];

        // inputs for 3 helpers
        let mut inputs = [vec![], vec![], vec![]];
        for row in data {
            let left = share(row.0, &mut rng);
            let right = share(row.0, &mut rng);

            for i in 0..3 {
                inputs[i].push((left[i], right[i]));
            }
        }

        let protocols: [_; 3] = inputs.iter()
            .map(|v| ToyProtocol { input: &v })
            .collect::<Vec<_>>().try_into().unwrap();

        let v = contexts.iter()
            .zip(protocols)
            .map(|(ctx, protocol)| protocol.execute(ctx))
            .collect::<FuturesOrdered<_>>()
            .collect::<Vec<_>>().await;
    }
}

```

This program fails with the following error

```bash

Refined 'protocol' with step 'toy_step_1' twice
thread 'protocol::tests::test' panicked at 'Refined 'protocol' with step 'toy_step_1' twice', src/protocol/mod.rs:92:13
stack backtrace:
```

The error occurs because we attempt to run the same circuit for more than one input row (record id). This should be supported by the infrastructure imo, because otherwise folks are forced to model protocols with unique steps created per row (see #108 and #110). Not only "artificial" steps look weird, they also make infra inefficient as all the batching benefits are gone (step per row effectively means a single record per channel). 

### Potential solutions

1. Move checks from `narrow` to `send`. `send` method has access to both step and record id, so it can validate that there is at most one send per step, per record, per query id. 
  * Pros: fixes the problem for message
  * Cons: `prss` method cannot benefit from this check, so we are still vulnerable to misuse of `prss` unless we implement something similar for it
2. Add `record_id` to `narrow` method. Enforce uniqueness of step + record_id rather than just step
  * Pros: fixes the issue for both `prss` and infra
  * Cons: it may feel unnatural to pass both record_id and step for composite protocols that operate on collections rather than single record. For example: shuffle wants to define a step that groups 3 sub-protocols, so it attempts to narrow the context but it needs to invent record_id out of thin air. 

any other ideas?

**Comment by martinthomson** [comment]
I had a discussion with @benjaminsavage about whether ProtocolContext should be passed by reference or by value.  This leads me to conclude that what we probably want here is a unique value for each record that is passed by value.

That sounds inefficient, and it currently is... at least a little.  But all we really need in the end is a single reference to the global state and a single integer value that captures the complete "Step" state.  In a debug, that might be bulky and inefficient (a String, plus other stuff), but a release build should be able to refine that down to a simple, small integer.  That means that ProtocolContext will effectively be able to be copied around all over.  A distinct copy for each record is totally reasonable in that context.

Note that we won't be able to implement Copy because the debug builds won't be so efficient, but that shouldn't be any real impediment as calling .clone() is fine.

**Comment by benjaminsavage** [comment]
That's fine, and I'm happy to call `.clone()` but we still need to move this check. 

I'm worried about option 1 because the PRSS stuff is precisely the place we are most likely to mess this up, and I don't want to miss it.

I wouldn't mind passing in record_id into calls to "narrow"; they're always in scope when narrow() gets called. So I vote for option 2

**Comment by akoshelev** [comment]
I agree with @benjaminsavage that protocol writers operate with records and steps and it feels intuitive. Certainly infra wants a unique "id" that identifies communication but to me it seems valuable if infra can provide a convenient abstraction over it. Maybe composing record id (optional?) and step in `narrow` is enough. Maybe fixing record_id in the context similarly to how we fix the step in narrow?

**Comment by martinthomson** [comment]
So I don't think that this needs to be a problem if we are careful with the API as it is currently defined.

The problem here occurs because a context is created for processing a list of records.  That process executes one or more functions on each record, passing a reference to that context each time.  The processing of the first record involves a call to `.narrow(X)`, and the same applies to each subsequent record.

To be clear, Alex's proposed fix totally works, but I don't think that it is necessary.

The alternative is to give the processing of each record a copy of the context.  Provided that we maintain discrete record identifiers associated with the processing of each record (which is necessary anyway), then we maintain separation.  The context, when narrowed by record 1 will be a different instance from the context that record 2 narrows.  Problem solved.

The only cost is that we are cloning the context a LOT.  For a debug build, that means cloning the de-duplication stuff, which might end up being a little expensive for larger datasets with lots of records, but debug builds shouldn't be processing large datasets.  Otherwise, we are building a tree (with strings) with a context instance at each node of the tree (including de-duplication check state) and replicating that structure for every record.  That cost is paid for every record that is actively being processed, with the overall cost being the number of records in progress times the depth of the tree (times the size of the structure).  Even in a debug build, I think that's manageable if we keep the number of in-progress records limited.

As for size... The context is currently four items: a role identifier, a reference to communications, a reference to PRSS state, and the step identifier.  The first three could easily be moved behind a single reference.  We already have a need to make the step identifier leaner (I think we can use a minimally-sized integer), so we can concentrate our efforts there.  If we can get down to less than 16 bytes, then creating copies is as close to free as possible.  We could use `Copy` then, but I'm concerned that that might produce accidental mistakes, so we'll stick to `Clone` in the mean time and have everything explicit.

Concrete steps: take the `ProtocolContext` instance by value everywhere.  I had one place in the shuffle code where a reference made sense, so that isn't a hard rule, but the public interface to your functions should take it by value.

**Comment by akoshelev** [comment]
Maybe I am missing something, but it looks like we are on the same page as we all agree that we need to narrow by record and step. The only difference is how to implement it. 

@martinthomson if I understand your proposal correctly, narrow by record must be done explicitly

```rust
 impl <'a, F: Field> ToyProtocol<'a, F> {
        async fn execute<N: Network>(self, ctx: ProtocolContext<'a, N>) -> Vec<Replicated<F>> {

            let mut futures = FuturesOrdered::new();

            // for every input row, execute `MultiplyTwice`
            for (record_id, r) in self.input.iter().enumerate() {
                let record_id = RecordId::from(record_id as u32);
                // explicitly narrow the context by record
                let ctx = ctx.narrow_to_record(record_id);
                let sub_protocol = MultiplyTwice { a: r.0, b: r.1, record_id };
                futures.push_back(sub_protocol.execute(ctx)); // context must be moved 
            }

            let res = futures.collect::<Vec<_>>().await;
            res
        }
    }
``` 

is this what you propose or I misunderstood you?

**Comment by martinthomson** [comment]
That would do it, yes.

**Comment by akoshelev** [comment]
I guess we only have one question unanswered before we can proceed with changing infra API which is: how we are going to ensure context is narrowed to a record only once

Implementing a runtime check would be fairly straightforward and cost us an extra 32 bits

```rust
   struct ProtocolContext {
      ...
      record_id: Option<RecordId>
   }

   impl ProtocolContext {
         fn narrow_to_record(self, record_id: RecordId) -> Self {
             if self.record_id.is_some() { panic!("...") }
             Self { ..., record_id: Some(record_id) }
         }
   }
```

compile-time check would be easy to implement, but harder to manage on the client side

```rust
   struct ProtocolContext {
     ...
   }
   struct RecordBoundProtocolContext {
      ...
      record_id: RecordId
   }

   impl ProtocolContext {
         fn narrow_to_record(self, record_id: RecordId) -> RecordBoundProtocolContext {
             NarrowedContext { ..., record_id: Some(record_id) }
         }
   }
```

every protocol would need to decide whether they should take `ProtocolContext` or `NarrowedProtocolContext` which should be fairly obvious. `Gateway` and `PrssEndpoint` will have to be taken away from the contexts (there seem to be another motivation for it in #139) and whatever new struct we use will need to be injected instead of `ctx` into multiplication and other essential protocols we use.

**Comment by benjaminsavage** [comment]
I vote for the latter. I like the idea of compile-time assistance. I think nearly all the protocols we are writing will accept a `NarrowedProtocolContext` with a few exceptions. But either way, we should know and be able to choose ourselves.

**Comment by akoshelev** [comment]
I spent some time looking into this issue again because it is blocking us from publishing working sort implementation. So there is a fairly straightforward way to bind `ProtocolContext` to records and have a runtime check. I think we could easily transform it to compile time if we need to. 

You can check the sketch of the implementation here https://github.com/richajaindce/raw-ipa/pull/5 

However there is one subtle issue that I don't really like. It is not directly caused by this change, we had it before too. 
Basically the same PRSS instance (backed by one private key) can be asked to generate random values at the same index. There is no runtime check for that.

Pseudo-rust code to illustrate the issue
```rust
   async fn protocol_1(ctx: ProtocolContext) {
        let randomness = ctx.prss().random_u128(0); // generate randomness at index 0;
        
        // create a sub-protocol and pass the context there
        protocol_2(ctx).await
   }

   async fn protocol_2(ctx: ProtocolContext) {
        let randomness = ctx.prss().random_u128(0); // generate randomness at the same index 0;
   }
```

Note that the issue may even be more subtle if multiplication is involved because there is nothing in the public API that tells the caller about how it uses `prss`. The fact that multiply uses record id to generate randomness is pretty much hidden behind its facade. We have some cases where record ids are invented inside the protocol https://github.com/private-attribution/ipa/blob/8834f6fc7c481b3e9a6714e1267f457d3ff596e4/src/protocol/attribution/accumulate_credit.rs#L179 so it is an easy to make mistake to use the same record id twice on the same PRSS. 

Note that the existing "strict" `narrow` behavior partially protects from this issue, but only in cases where `narrow` is used. The example above would cause the same issue even today.

**Comment by andyleiserson** [comment]
I agree that a safety check on the PRSS is a good idea. Something like a check that `GeneratorFactory::generate` is not called twice with the same context (or else changes to the API to make it harder to do so).

I do think that binding by record ID is necessary, but I'm not sure it resolves this issue unless we track each record separately in the `used` `HashMap`, which will have a pretty high cost. I think we have what is effectively a netlisting problem here -- it's not okay to have two gates in the design with the same hierarchical path. The current code assembles and checks the uniqueness of the hierarchical paths in the inner loop of execution. In other words, `ctx.narrow` is saying "instantiate a gate at this path, throw an error if that path is not unique". Which is the correct logic if you're assembling a description of the circuit that can be reused across multiple executions (a netlist, in HW speak), but not correct logic in the actual evaluation of the gate.

One way to deal with this is to move `ctx.narrow` from `execute` into `new`, but this isn't consistent with the current use of `ConvertShares::new` to pass in the actual share bits for a particular execution. I think maybe the share bits should be an argument to `ConvertShares::execute` rather than `ConvertShares::new`.

The existing `gen_random` test looks like it might be exercising this case, because it does `RecordId::from(i)`, but I don't think it really should be doing that -- `i` is iterating bits, not records. If I change it to do two 40-bit values (for two records), it fails the same way as `convert_one_bit_of_many_match_keys`.

**Comment by andyleiserson** [comment]
To be more concrete what I mean by "move ctx.narrow from execute into new": https://github.com/private-attribution/ipa/compare/main...andyleiserson:ipa:andy/issue135

**Comment by akoshelev** [comment]
>  do think that binding by record ID is necessary, but I'm not sure it resolves this issue unless we track each record separately in the used HashMap, which will have a pretty high cost.

Completely agree here, it does not resolve this issue at all. Just makes it more visible imo

> To be more concrete what I mean by "move ctx.narrow from execute into new"

This problem seems like something that protocols shouldn't care about as long as their code compiles and does not panic in runtime.  if we can find a way to prevent protocols from narrowing the context if it is not possible and at the same time allow narrowing more than once if it is correct (when narrow happens per record) that would be the ideal solution I think

**Comment by andyleiserson** [comment]
Another line of thought I had was to allow narrowing the same context multiple times as long as it's done at a single source location. Enforcing that could be done by remembering the file and line number, or maybe by remembering something in a `static OnceCell`.

I'll think about it some more. I agree it would be nice to have a single high-level description of the protocol rather than splitting it into two pieces like I did in my branch. I'm just not sure the best way of doing that.

**Comment by andyleiserson** [comment]
Here is another version, using `OnceCell`s: https://github.com/private-attribution/ipa/compare/main...andyleiserson:ipa:andy/issue135-v3a?expand=1

**Comment by andyleiserson** [comment]
Fixed within https://github.com/private-attribution/ipa/pull/253

**Comment by akoshelev** [comment]
oh, right, thanks!
